runtime/dynstats.c

```diff
@@ -499,12 +643,161 @@ dynstats_createCtr(dynstats_bucket_t *b, const uchar* metric, dynstats_ctr_t **c
+{
+	int lenout;
+	const uchar* pszworkdir;
+
+	/* Get Raw Workdir, if it is NULL we need to propper handle it */
+	pszworkdir = getStateFileDir(b);
+
+	/* Construct file name */
+	lenout = snprintf((char*)pszout, buflen, "%s/%s",
+					(char*) (pszworkdir == NULL ? "." : (char*) pszworkdir), (char*)pszstatefile);
+
+	/* return out length */
+	return lenout;
+}
+
+/* this generates a state file name suitable for the given file. To avoid
+ * malloc calls, it must be passed a buffer which should be MAXFNAME large.
+ * Note: the buffer is not necessarily populated ... always ONLY use the
+ * RETURN VALUE!
+ * This function is guranteed to work only on config data and DOES NOT
+ * open or otherwise modify disk file state.
+ */
+static uchar * ATTR_NONNULL(1, 2)
+getStateFileName(const dynstats_bucket_t *const pbucket,
+		uchar *const __restrict__ buf,
+		const size_t lenbuf) {
+
+	DBGPRINTF("getStateFileName for '%s'\n", pbucket->name);
+	snprintf((char*)buf, lenbuf - 1, "dynstats-state:%s", pbucket->name);
+	DBGPRINTF("getStateFileName:  state file name now is %s\n", buf);
+	return buf;
+}
+
+/* This function persists dynstats_bucket_t data, which for the time being:
+ * metric bucket counters
+ */
+static rsRetVal ATTR_NONNULL(1)
+persistBucketState(dynstats_bucket_t *b) {
+	DEFiRet;
+	uchar statefile[MAXFNAME];
+	uchar statefname[MAXFNAME];
+	struct json_object *jval = NULL;
+	struct json_object *json = NULL;
+	struct json_object *json_bucket_values = NULL;
+
+	assert (b->name);
+	uchar *const statefn = getStateFileName(b, statefile, sizeof(statefile));
+	/* Get full path and file name */
+	getFullStateFileName(b, statefn, statefname, sizeof(statefname));
+	DBGPRINTF("persisting state for '%s', state file '%s'\n", b->name, statefname);
+
+	CHKmalloc(json = json_object_new_object());
+	jval = json_object_new_string((char*) b->name);
+	json_object_object_add(json, DYNSTATS_BUCKET_PERSIST_NAME, jval);
+
+	// walk the hashtable and create a val for each
+	CHKmalloc(json_bucket_values = json_object_new_object());
+	json_object_object_add(json, DYNSTATS_BUCKET_PERSIST_VALUES_NAME, json_bucket_values);
+
+	if (hashtable_count(b->table) > 0) {
+		struct hashtable_itr *itr = hashtable_iterator(b->table);
+		dynstats_ctr_t *pctr = NULL;
+		do {
+			pctr = (dynstats_ctr_t*) hashtable_iterator_value(itr);
+			jval = json_object_new_int64(pctr->ctr);
+			json_object_object_add(json_bucket_values, (const char*)pctr->metric, jval);
+		} while (hashtable_iterator_advance(itr));
+		free(itr);
+	}
+
+	const char *jstr =  json_object_to_json_string_ext(json, JSON_C_TO_STRING_PLAIN);
+	CHKiRet(enqueueFileWriteTask(b, (const char*)statefname, jstr));
+
+finalize_it:
+	if(iRet != RS_RET_OK) {
+		LogError(0, iRet, "dynstats: could not persist state "
+				"file %s - data may be repeated on next "
+				"startup. Is WorkDirectory set?",
+				statefname);
+	}
+	json_object_put(json);
```

Comment:
This looks like invalid memory access.

Given `jstr` above is `const char*` I presume it is tied to `json`.
And you just free'ed `json` here. So worker would end up performing illegal memory access when it reads `content`, won't it?

Reply:
since we just copy the json content string to the task structure, that is why we immediately free it here. So no need to keep it around.

Reply:
Yes, sorry about the false alarm.